tsdb/querier.go

```diff
@@ -376,6 +378,34 @@ func labelValuesWithMatchers(r IndexReader, name string, matchers ...*labels.Mat
 		allValues = filteredValues
 	}
 
+	// Let's see if expanded postings for matchers have smaller cardinality than label values.
+	// Since computing label values from series is expensive, we apply a limit on number of expanded
+	// postings (and series).
+	maxExpandedPostings := len(allValues) / maxExpandedPostingsFactor
+	if maxExpandedPostings > 0 {
+		// Add space for one extra posting when checking if we expanded all postings.
+		expanded := make([]storage.SeriesRef, 0, maxExpandedPostings+1)
+
+		for len(expanded) < maxExpandedPostings && p.Next() {
+			expanded = append(expanded, p.At())
+		}
+
+		// Did we reach end of postings?
+		reachedEnd := true
+		if len(expanded) == maxExpandedPostings && p.Next() {
+			reachedEnd = false
+			expanded = append(expanded, p.At())
+		}
+
+		if reachedEnd {
+			// We have expanded all the postings -- all returned label values will be from these series only.
+			return labelValuesFromSeries(r, name, expanded)
+		}
```

Comment:
Also, can we just do:
```suggestion
		reachedEnd := true
		if len(expanded) < maxExpandedPostings { // Implies that `p.Next()` was false last time we called it.
			if p.Err() != nil {
				return nil, p.Err()
			}
			return labelValuesFromSeries(r, name, expanded)
		}
```

Reply:
We could, although that would allow for one less expanded postings than our computed `max`. I'll keep the current version, but add missing `p.Err()` (thanks for catching that!).

Reply:
True, but we could fix that by changing both `for` and `if` to match less or equal instead of strictly less. I just find it strange and unnecessary to have a separate "add one more" case for the last posting (with an extra bool to indicate we didn't find it), instead of using the condition of the for loop that we've done already.